crates/bevy_render/src/render_phase/mod.rs

```diff
@@ -51,18 +52,227 @@ use std::{ops::Range, slice::SliceIndex};
+            }
+        }
+    }
+
+    /// Executes the GPU commands needed to render all entities in this phase.
+    pub fn render<'w>(
+        &self,
+        render_pass: &mut TrackedRenderPass<'w>,
+        world: &'w World,
+        view: Entity,
+    ) {
+        let draw_functions = world.resource::<DrawFunctions<BPI>>();
+        let mut draw_functions = draw_functions.write();
+        draw_functions.prepare(world);
+
+        // Draw batchables. Track instance index and dynamic offset, if
+        // applicable. We assume that all instance indices and dynamic offsets
+        // are consecutive.
+
+        let mut first_instance_index = self.first_instance_index;
+        let mut first_dynamic_offset = self.first_dynamic_offset;
+
+        debug_assert_eq!(self.batchable_keys.len(), self.batches.len());
+
+        for (key, batch) in self.batchable_keys.iter().zip(self.batches.iter()) {
+            let binned_phase_item = BPI::new(
+                key.clone(),
+                batch.representative_entity,
+                first_instance_index..batch.last_instance_index,
+                first_dynamic_offset,
+            );
+
+            let draw_function = draw_functions
+                .get_mut(binned_phase_item.draw_function())
+                .unwrap();
+            draw_function.draw(world, render_pass, view, &binned_phase_item);
+
+            // Advance dynamic offset and instance index.
+            if let Some(dynamic_offset) = first_dynamic_offset {
+                first_dynamic_offset = (u32::from(dynamic_offset) + batch.last_instance_index
+                    - first_instance_index)
+                    .try_into()
+                    .ok();
+            }
+            first_instance_index = batch.last_instance_index;
+        }
+
+        // Draw unbatchables, which immediately follow the batchables in the
+        // storage buffers.
+
+        for key in &self.unbatchable_keys {
+            for &entity in &self.unbatchable_values[key] {
+                let binned_phase_item = BPI::new(
+                    key.clone(),
+                    entity,
+                    first_instance_index..(first_instance_index + 1),
+                    first_dynamic_offset,
+                );
+
+                let draw_function = draw_functions
+                    .get_mut(binned_phase_item.draw_function())
+                    .unwrap();
+                draw_function.draw(world, render_pass, view, &binned_phase_item);
+
+                // Advance dynamic offset and instance index.
+                first_instance_index += 1;
+                if let Some(dynamic_offset) = first_dynamic_offset {
+                    first_dynamic_offset = (u32::from(dynamic_offset) + 1).try_into().ok();
+                }
+            }
+        }
+    }
+}
+
+impl<BPI> Default for BinnedRenderPhase<BPI>
+where
+    BPI: BinnedPhaseItem,
+{
+    fn default() -> Self {
+        Self {
+            batchable_keys: default(),
```

Comment:
I'm not sure what this means.

Reply:
Vec::new() and the other equivalent constructors can be const, but Default::default need to at least be inlined in some way. It may be preferable to use const construction where possible.